### DISCUSSION TITLE
Direction of the collision between rectangles

### DISCUSSION THREAD

**Discussion by ahmedatef286** [raysan5/raylib#1983] [Q&A] [answered]
Hi everyone hope you're having a fun time with raylib as i am! I'm quite the beginner and don't know if the solution is too obvious and I'm just missing it
Is there an efficient way to detect the direction of collision in terms of one rectangle in relation to the other? For example i have a collision rectangle representing a character in a 2d environment and want to know when it collides with another rectangle is the collision happening to the right/left/up/bottom of the character , what I've been doing is comparing the coordinates and just winging it but i don't know if that's the best way since it keeps requiring tuning

**Comment by jmorel33** [comment, marked as answer]
Once you test for collision, just check a.x < b.x where character a's x position would be more left than character b.  Super easy in 2D.  If you said 3D, you'd have to think of a 3'rd axis.  That is my brute force way which always works essentially.

recap;
1) do you have collision
2) if, so, which character is more, left or more right, and by how much

If you need to do collision detection for different places, you would have to have multiple hit points on your character and detect which one it was.  I wouldn't even use triangle, but a radius. or a box

**Reply by ahmedatef286** [reply]
Thanks! 
I did end up implement that tactic but probably in a less efficient way , I'm learning so i didn't have velocity in mind while i was implementing it , so if my character [ rectangle] was moving fast enough it ended up affecting the direction I'm resolving the collision towards.
Also i implemented gravity just by constantly moving my character downwards and switching it off when it get in collision with something under it but it ended up preventing my character from walking off ledges 😤
I found a really good tutorial on YouTube about rectangle collision using 2d rays (https://youtu.be/8JJ-4JgR7Dg     if someone in the future needing it ) ,which takes in consideration the expected time you'll collide with another object and uses it to determine how quick you need to resolve it as soon as collision is finally established.
Also learned about clamping  my velocity in all direction to a maximum to further smooth things out and it works like a treat now.


Currently I'm facing a problem i tried adding a condition for climbing over ledges (if a.y is greater than b.y and a.x + a.height is <= b.y+b.height) my character should start moving upwards 
In my implementation i think i messed up the order of resolving collisions and allowing my character to climb over a ledge because currently while my character is in Climbing mode it stops abiding to the collision detection rules i made ( interestingly only vertically , horizontal collision seems fine) 
I will add all the code after making it look a bit more decent and less like a Isaac Newton discovering gravity , it is very amateurish ofc but would be nice for someone still learning like myself. Thanks again!!!

**Reply by jmorel33** [reply]
2D collision imho is really easy if you consider you could use the bitmap as collision detection, so if one pixel of your character is above a non null pixel (may choose in RGB I the I = 0 value as a means to say void nothing to collide with, or you may choose combined RGB > 0.  I do not think checking a group of pixels is necessarily slow, unless you check thousands every frame it becomes intense at some point.

The other better way I find is checking background or even foreground collision with a character map or tile map, that way if the tile is non zero it is instant collision...
